src/main/java/net/neoforged/neoforge/energy/ItemEnergyStorage.java

```diff
@@ -0,0 +1,113 @@
+    protected final int maxExtract;
+
+    /**
+     * Creates a new ItemEnergyStorage with a data component as the backing store for energy capacity.
+     * 
+     * @param parent         The parent item stack
+     * @param powerComponent The data component referencing the stored energy of the item stack
+     * @param capacity       The max capacity of the energy being stored
+     * @param maxReceive     The max per-transfer power input rate
+     * @param maxExtract     The max per-transfer power output rate
+     */
+    public ItemEnergyStorage(ItemStack parent, DataComponentType<Integer> powerComponent, int capacity, int maxReceive, int maxExtract) {
+        this.parent = parent;
+        this.powerComponent = powerComponent;
+        this.capacity = capacity;
+        this.maxReceive = maxReceive;
+        this.maxExtract = maxExtract;
+    }
+
+    /**
+     * Creates a new ItemEnergyStorage with a unified receive / extract rate.
+     * 
+     * @see ItemEnergyStorage#ItemEnergyStorage(ItemStack, DataComponentType, int, int, int)
+     */
+    public ItemEnergyStorage(ItemStack parent, DataComponentType<Integer> powerComponent, int capacity, int maxTransfer) {
+        this(parent, powerComponent, capacity, maxTransfer, maxTransfer);
+    }
+
+    /**
+     * Creates a new ItemEnergyStorage with a transfer rate equivalent to the capacity.
+     * 
+     * @see ItemEnergyStorage#ItemEnergyStorage(ItemStack, DataComponentType, int, int, int)
+     */
+    public ItemEnergyStorage(ItemStack parent, DataComponentType<Integer> powerComponent, int capacity) {
+        this(parent, powerComponent, capacity, capacity);
+    }
+
+    @Override
+    public int receiveEnergy(int maxReceive, boolean simulate) {
+        if (!canReceive())
+            return 0;
+
+        int energy = this.getEnergyStored();
+        int energyReceived = Math.min(capacity - energy, Math.min(this.maxReceive, maxReceive));
+        if (!simulate) {
+            this.setEnergy(energy + energyReceived);
+        }
+        return energyReceived;
+    }
+
+    @Override
+    public int extractEnergy(int maxExtract, boolean simulate) {
+        if (!canExtract())
+            return 0;
+
+        int energy = this.getEnergyStored();
+        int energyExtracted = Math.min(energy, Math.min(this.maxExtract, maxExtract));
+        if (!simulate) {
+            this.setEnergy(energy - energyExtracted);
+        }
+        return energyExtracted;
+    }
+
+    @Override
+    public int getEnergyStored() {
+        int rawEnergy = this.parent.getOrDefault(this.powerComponent, 0);
+        return Math.max(0, Math.min(capacity, rawEnergy));
+    }
+
+    @Override
+    public int getMaxEnergyStored() {
+        return this.capacity;
+    }
+
+    @Override
+    public boolean canExtract() {
+        return this.maxExtract > 0;
+    }
+
+    @Override
+    public boolean canReceive() {
+        return this.maxReceive > 0;
+    }
+
+    protected void setEnergy(int energy) {
```

Comment:
Do we want to delete the component when it goes down to 0? Maybe not.

Reply:
Probably not, no.  I think that would be confusing against typical user expectations (and is probably in contrast of how components are normally used)